make bookshelf power configurable and refresh enchantment hint of menu after editing bookshelf power

I-am-ddang opened the pull request.

Strokkur424 reviewed (changes requested):
What I find a bit unfortunate is that changing the bonus power in the event basically makes any changes in the offers redundant, as you reroll these anyways.

I don't know if there is a particularly clean way to fix this. I personally can also only think of one way to fix this, but I understand it is not very advantageous:

- The introduction of a new `RetrieveItemEnchantPowerEvent`, which allows for setting the enchantment bonus before the offers are even put into consideration.

Therefore, the actual PrepareItemEnchantEvent could stay unmodified and enchantment offer modifications would still work as intended.

Of course, before you make this change, I'd like a Paper developer to confirm that this is an acceptable step to take, because I am not sure myself either.

lynxplay committed:
Roll EnchantmentOffer API

Introduces a new API method that allows plugins to simulate the rolling
process of an enchantment table for enchantment offers. This is
specifically useful to "reroll" an enchantment tables offers during the
PrepareItemEnchantEvent to e.g. change the bookshelf bonus or seed.

Co-authored-by: I-am-DDang_ <psychologist3552@gmail.com>

I-am-ddang committed:
make bookshelf power configurable and refresh enchantment hint of menu after editing bookshelf power

lynxplay committed:
Revert "make bookshelf power configurable and refresh enchantment hint of menu after editing bookshelf power"

This reverts commit b56273bef901949737986353ea95586f02f06db3.

lynxplay committed:
Missing annotations

lynxplay committed:
Merge remote-tracking branch 'origin/main' into feat-enchantment-hint-with-configurable-bookshelf

lynxplay committed:
Merge remote-tracking branch 'origin/main' into feat-enchantment-hint-with-configurable-bookshelf
